### PR TITLE
cv::TickMeter class addition

### LINKED ISSUE #6582: Feature Request cv::TickMeter
i think adding cv::TickMeter class in the library will be fine.

there is open questions at answers.opencv.org
http://answers.opencv.org/question/67009/why-tickmeter-class-only-used-in-the-gpu-samples/
http://answers.opencv.org/question/69035/tickmeter-class-removal-and-how-to-replace-it/

PS: if some core developer confirm and suggest the proper module ( i think it can be merged in core module). i will try to supply Pull Request.

### PR DESCRIPTION
**Pull request by sturkmen72**
resolves #6582

### What does this PR change?
moves cv::TickMeter class in the core library

initial commit

### CODE REVIEW

**Comment by paroj** [comment]
never really missed that class..

But I think it could be more useful if it allowed computing the average time over multiple runs instead of just
```cpp
auto start = cv::getTickCount();
...
(cv::getTickCount() - start)/cv::getTickFrequency();
```

**Comment by sturkmen72** [comment]
@paroj actually it allows to calculate the average time over multiple runs

```

#include "opencv2/imgproc.hpp"
#include "opencv2/highgui.hpp"
#include <iostream>

using namespace cv;
using namespace std;

int main()
{
	TickMeter tm;
	Mat test(1000, 1000, CV_8U);
	randu(test, 0, 255);
	Mat thresh;
	cout << "thresholding a 1000x1000 Mat 5000 times\n\n";
	for (int i = 0; i < 5000; i++)
	{
		tm.start();
		threshold(test, thresh, 230, 255, THRESH_BINARY);
		tm.stop();
	}

	cout << "Total Time in microseconds :" << tm.getTimeMicro() << endl;
	cout << "Average Time in microseconds :" << tm.getTimeMicro() / tm.getCounter() << endl;

	cout << "Total Time in milliseconds :" << tm.getTimeMilli() << endl;
	cout << "Average Time in milliseconds :" << tm.getTimeMilli() / tm.getCounter() << endl;

	cout << "Total Time in seconds :" << tm.getTimeSec() << endl;
	cout << "Average Time in seconds :" << tm.getTimeSec() / tm.getCounter() << endl;

	imshow("thresh", thresh);
	waitKey();
    return 0;
}
```
output:

thresholding a 1000x1000 Mat 5000 times

Total Time in microseconds :317603
Average Time in microseconds :63.5205
Total Time in milliseconds :317.603
Average Time in milliseconds :0.0635205
Total Time in seconds :0.317603
Average Time in seconds :6.35205e-05

**Comment by StevenPuttemans** [comment]
Hello @sturkmen72 in relation to issue #6502, do you know if the tickMeter reports the same time in both windows and linux systems? I am looking for a universal way of providing time, even if multicore processing is being used.

**Comment by sturkmen72** [comment]
@StevenPuttemans Sorry i have no idea about linux systems :(

File: modules/core/include/opencv2/core/utility.hpp [hunk 1]
```diff
@@ -264,9 +265,53 @@ execution time in seconds:
...
+    resets internal values.
+    */
+    CV_WRAP void reset();
+private:
+    int64 counter;
+    int64 sumTime;
+    int64 startTime;
+};
+
+std::ostream& operator << (std::ostream& out, const TickMeter& tm);
```
Comment (sturkmen72) [inline modules/core/include/opencv2/core/utility.hpp]: i need some help about documentation. how can we give an example usage of << operator ?

File: modules/core/include/opencv2/core/utility.hpp [hunk 2]
```diff
@@ -264,9 +265,53 @@ execution time in seconds:
...
+tm.stop();
+std::cout << tm.getTimeSec();
+@endcode
+@sa getTickCount, getTickFrequency
+*/
+
+class CV_EXPORTS_W TickMeter
+{
+public:
+    //! the default constructor
```
Comment (sturkmen72) [inline modules/core/include/opencv2/core/utility.hpp]: any help about explanation of methods is appreciated

**Comment by FredFG** [comment]
Time is tricky. It is so critical to modern computing there's an ISO standard: iso.20.11.7
C++ implements this standard.
TickMeter is in OpenCv version 2.4 and therefore is C based. I wouldn't use it.
It should be deprecated.

Do your own timing with C++ methods in std::chrono and you should get comparable results on any modern OS or computer. 

A simple example from Stroustrup The C++ Programming Language, 4th edition. Section 35.2.

steady_clock_time::time_point t = steady_clock::now();
// do your stuff
steady_clock::duration d1 = steady_clock::now() - t;
// do some more stuff
steady_clock::duration d2 = steady_clock::now() - d1;

There's nothing easier than that.
You can get resolutions down to picoseconds (depending on your hardware). It doesn't care how many cpus are running or what you had for breakfast.

Really precise timing might get weird on a "really cheap" CPU such as Raspberry Pi or Arduino. Even these errors would be so small they would likely be meaningless unless you're an astrophysicist.

**Comment by StevenPuttemans** [comment]
I kind of agree with @FredFG that the absence of ISO standards for time and the introduction of them using C++11 can help us to make a more universal approach. However, C++11 is far from the standard compiler for now, so maybe we should add the 2 cases
1. if c++11 is available use new functionality
2. else fall back

File: modules/core/src/system.cpp [hunk 3]
```diff
@@ -508,6 +508,27 @@ int64 getCPUTickCount(void)
...
+{
+    int64 time = cv::getTickCount();
+    if (startTime == 0)
+    return;
+    ++counter;
+    sumTime += (time - startTime);
+    startTime = 0;
+}
+
+std::ostream& operator << (std::ostream& out, const TickMeter& tm) { return out << tm.getTimeSec() << "sec"; }
```
Comment (alalek) [inline modules/core/src/system.cpp]: This implementation is very light, so I suggest to put all methods it into header directly (if bindings generator is fine to bypass this code).
Especially for ostream-based `operator <<`

File: modules/core/src/system.cpp [hunk 4]
```diff
@@ -508,6 +508,11 @@ int64 getCPUTickCount(void)
 
 #endif
 
+std::ostream& operator << (std::ostream& out, const TickMeter& tm)
```
Comment (sturkmen72) [inline modules/core/src/system.cpp]: @alalek i tried to move this part in utility.hpp but get errors. what do you suggest?

Comment (alalek) [inline modules/core/src/system.cpp]: What kind of errors do you have? Did you try "static inline" without CV_EXPORTS?

Comment (sturkmen72) [inline modules/core/src/system.cpp]: @alalek thanks. it is OK with "static inline"

**Comment by vpisarev** [comment]
this PR will be recreated because of CI issues

**Pull request closed without merging**